From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Mon, 21 Aug 2023 21:46:10 +0200
Subject: [PATCH] Pre-compute VarLong sizes

License: GPL-3.0-only (https://www.gnu.org/licenses/gpl-3.0.html)
Gale - https://galemc.org

Based on Velocity's varint/varlong writing optimization. Packet encoding calls
VarLong#getByteSize frequently; use a tiny immutable lookup table while keeping
a folia.yml escape hatch to the original loop.


diff --git a/net/minecraft/network/VarLong.java b/net/minecraft/network/VarLong.java
--- a/net/minecraft/network/VarLong.java
+++ b/net/minecraft/network/VarLong.java
@@ -9,6 +9,22 @@ public class VarLong {
     private static final int DATA_BITS_PER_BYTE = 7;
 
     public static int getByteSize(final long value) {
+        if (dev.eren.folia.config.FoliaConfig.precomputeVarLongSizes) {
+            return VARLONG_EXACT_BYTE_LENGTHS[Long.numberOfLeadingZeros(value)];
+        }
+        return getByteSizeOld(value);
+    }
+
+    private static final int[] VARLONG_EXACT_BYTE_LENGTHS = new int[65];
+    static {
+        for (int i = 0; i < 64; i++) {
+            VARLONG_EXACT_BYTE_LENGTHS[i] = (int)Math.ceil((63d - (i - 1)) / 7d);
+        }
+        VARLONG_EXACT_BYTE_LENGTHS[64] = 1;
+    }
+
+    static int getByteSizeOld(final long value) {
         for (int i = 1; i < MAX_VARLONG_SIZE; i++) {
             if ((value & -1L << i * DATA_BITS_PER_BYTE) == 0L) {
                 return i;
